core/src/main/java/io/questdb/cutlass/http/processors/IlpTudCache.java

```diff
@@ -0,0 +1,231 @@
+
+    @Override
+    public void close() {
+        // Close happens when HTTP connection is closed
+        ObjList<Utf8String> keys = tableUpdateDetails.keys();
+        for (int i = 0, n = keys.size(); i < n; i++) {
+            Utf8Sequence tableName = tableUpdateDetails.keys().get(i);
+            WalTableUpdateDetails tud = tableUpdateDetails.get(tableName);
+            Misc.free(tud);
+        }
+        tableUpdateDetails.clear();
+        Misc.free(ddlMem);
+        Misc.free(path);
+    }
+
+    public void commitAll() throws Throwable {
+        boolean droppedTableFound;
+        do {
+            droppedTableFound = false;
+            ObjList<Utf8String> keys = tableUpdateDetails.keys();
+            for (int i = 0, n = keys.size(); i < n; i++) {
+                Utf8Sequence tableName = tableUpdateDetails.keys().get(i);
+                WalTableUpdateDetails tud = tableUpdateDetails.get(tableName);
+                try {
+                    if (!tud.isDropped()) {
+                        tud.commit(false);
+                    }
+                } catch (CommitFailedException e) {
+                    // We can ignore dropped tables and simply roll back the transaction
+                    if (!e.isTableDropped()) {
+                        // This real commit error, abort the batch commit
+                        throw e.getReason();
+                    } else {
+                        // Remove dropped table TUD from cache
+                        // and repeat the commit procedure
+                        tud.setIsDropped();
+                    }
+                }
+
+                if (tud.isDropped()) {
+                    tableUpdateDetails.remove(tableName);
+                    Misc.free(tud);
+                    droppedTableFound = true;
+                    break;
+                }
+            }
+        } while (droppedTableFound);
+    }
+
+    public WalTableUpdateDetails getTableUpdateDetails(
+            SecurityContext securityContext,
+            @NotNull LineTcpParser parser,
+            Pool<SymbolCache> symbolCachePool
+    ) throws TableCreateException {
+        int key = tableUpdateDetails.keyIndex(parser.getMeasurementName());
+        if (key < 0) {
+            return tableUpdateDetails.valueAt(key);
+        }
+
+        tableNameUtf16.clear();
+        Utf8s.utf8ToUtf16(parser.getMeasurementName(), tableNameUtf16);
+        TableToken tableToken = getOrCreateTable(securityContext, parser, tableNameUtf16);
+        if (!engine.isWalTable(tableToken)) {
+            throw parseException.of("cannot insert in non-WAL table", null);
+        }
+
+        TelemetryTask.store(telemetry, TelemetryOrigin.ILP_TCP, TelemetrySystemEvent.ILP_RESERVE_WRITER);
+        // check if table on disk is WAL
+        path.of(engine.getConfiguration().getRoot());
+        Utf8String nameUtf8 = Utf8String.newInstance(parser.getMeasurementName());
+        WalTableUpdateDetails tud = new WalTableUpdateDetails(
```

Comment:
we create a new `tud` instance for each request. (since this cache is cleared upon parsin headers) that could be a lot of garbage. is it expected? is it similar in WAL-ed TCP ILP?

Reply:
At the moment I cache TUDs per TCP connection and clear on TCP disconnect. Http request boundary calls `cache.clear()` and TCP disconnect uses `cache.reset()`.

This is not ideal either, so I'm open to suggestions

Reply:
Better strategy probably would be to keep TUDs per TCP connection but release WAL writers on HTTP connection close back to WAL Writer pool. This will not leak many WAL Writers which are heavy in case of HTTP TCP connection pools on the client. What do you think?

Reply:
I missed the distinction between `cache.clear()` and `cache.reset()`, my bad, sorry about confusion. it's not as bad as I thought. 

to your question: my gut tells me that returning WAL Writers on HTTP connection boundaries is an unnecessary complication (and possibly also a small perf. penalty) because clients won't keep idle connections in pool for long (as discussed y'day)